File: 13-2.rb
#!/usr/bin/env ruby

def compare(p1, p2)
  left = p1[0]
  right = p2[0]

  return 0 if left.nil? and right.nil?
  return -1 if left.nil?
  return 1 if right.nil?

  if left.is_a? Integer and right.is_a? Integer
    if left == right
      compare(p1[1..], p2[1..])
    elsif left < right
      -1
    else
      1
    end
  elsif left.is_a? Array and right.is_a? Array
    case compare(left, right)
    when 0
      compare(p1[1..], p2[1..])
    when -1
      -1
    when 1
      1
    end
  elsif left.is_a? Integer
    case compare([left], right)
    when 0
      compare(p1[1..], p2[1..])
    when -1
      -1
    when 1
      1
    end
  elsif right.is_a? Integer
    case compare(left, [right])
    when 0
      compare(p1[1..], p2[1..])
    when -1
      -1
    when 1
      1
    end
  else
    raise "Type error: #{left}, #{right}"
  end
end

dividers = [[[2]], [[6]]]

input = File.read('13.input').lines.map(&:strip).map { |l| eval(l) }
input.reject! &:nil?
input += dividers

input.sort! { |a, b| compare(a, b) }

print dividers.map { |d| input.index(d) + 1 }.reduce(:*), "\n"

File: 13-2.py
#!/usr/bin/env python

from functools import cmp_to_key


def compare(p1, p2):
    if len(p1) == 0 and len(p2) == 0:
        return 0
    if len(p1) == 0:
        return -1
    if len(p2) == 0:
        return 1

    left = p1[0]
    right = p2[0]

    if type(left) is int and type(right) is int:
        if left == right:
            return compare(p1[1:], p2[1:])
        elif left < right:
            return -1
        else:
            return 1
    elif type(left) is list and type(right) is list:
        r = compare(left, right)
        if r == 0:
            return compare(p1[1:], p2[1:])
        else:
            return r
    elif type(left) is int:
        r = compare([left], right)
        if r == 0:
            return compare(p1[1:], p2[1:])
        else:
            return r
    elif type(right) is int:
        r = compare(left, [right])
        if r == 0:
            return compare(p1[1:], p2[1:])
        else:
            return r
    else:
        raise RuntimeError(f"Type error: {left}, {right}")


dividers = [[[2]], [[6]]]

input = [eval(row.strip()) for row in open('13.input').readlines() if row != "\n"]
input += dividers

s = sorted(input, key=cmp_to_key(compare))

key = 1
for d in dividers:
    key *= s.index(d) + 1

print(key)
